man/IntersectTool.1
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man v6.0.2 (Pod::Simple 3.45)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Required to disable full justification in groff 1.23.0.
.if n .ds AD l
.\" ========================================================================
.\"
.IX Title "INTERSECTTOOL 1"
.TH INTERSECTTOOL 1 2025-11-05 "GeographicLib 2.7" "GeographicLib Utilities"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
IntersectTool \-\- find the intersection of geodesics
.SH SYNOPSIS
.IX Header "SYNOPSIS"
\&\fBIntersectTool\fR [ \fB\-c\fR | \fB\-n\fR | \fB\-i\fR | \fB\-o\fR |
[ \fB\-R\fR \fImaxdist\fR ]
[ \fB\-e\fR \fIa\fR \fIf\fR] [ \fB\-E\fR ]
[ \fB\-w\fR ] [ \fB\-p\fR \fIprec\fR ]
[ \fB\-\-comment\-delimiter\fR \fIcommentdelim\fR ]
[ \fB\-\-version\fR | \fB\-h\fR | \fB\-\-help\fR ]
[ \fB\-\-input\-file\fR \fIinfile\fR | \fB\-\-input\-string\fR \fIinstring\fR ]
[ \fB\-\-line\-separator\fR \fIlinesep\fR ]
[ \fB\-\-output\-file\fR \fIoutfile\fR ]
.SH DESCRIPTION
.IX Header "DESCRIPTION"
\&\fBIntersectTool\fR finds the intersection of two geodesics \fIX\fR and
\&\fIY\fR.  The geodesics may either be specified as a location and an
azimuth, \fIlatX\fR \fIlonX\fR \fIaziX\fR, or as the shortest geodesic segment
between two locations, \fIlatX1\fR \fIlonX1\fR and \fIlatX2\fR \fIlonX2\fR.  The
intersection is then specified as the displacements, \fIx\fR and \fIy\fR,
along the geodesics \fIX\fR and \fIY\fR from the starting points to the
intersection.  In the case of the intersection of geodesic segments,
the starting point is first point specified for \fIX\fR or \fIY\fR.
.PP
Usually this tool returns the \fIclosest\fR intersection defined as the
one that minimizes the "L1" distance, |\fIx\fR| + |\fIy\fR|.  However, it is
possible to specify an "origin" \fIx0\fR and \fIy0\fR when determining
closeness so that the intersection which minimizes |\fIx\fR \- \fIx0\fR| +
|\fIy\fR \- \fIy0\fR| is returned.
.PP
In the case of intersecting segments the origin is taken to be the
midpoints of the segments; \fIx0\fR is half the distance from \fIX1\fR to
\&\fIX2\fR.  In addition a flag is returned specifying whether the
intersection is "within" the segments.
.PP
The tool also returns a "coincidence indicator" \fIc\fR.  This is
typically 0.  However if the geodesics lie on top of one another at
the point of intersection, then \fIc\fR is set to 1, if they are
parallel, and \-1, if they are antiparallel.
.PP
\&\fBIntersectTool\fR operates in one of three modes:
.IP 1. 4
With the \fB\-c\fR option (the default), \fBIntersectTool\fR accepts lines on
the standard input containing \fIlatX\fR \fIlonX\fR \fIaziX\fR \fIlatY\fR \fIlonY\fR
\&\fIaziY\fR, specifying two geodesic lines \fIX\fR and \fIY\fR, and prints the
location of the closest intersection \fIx\fR \fIy\fR \fIc\fR on standard output.
.IP 2. 4
With the \fB\-n\fR option, \fBIntersectTool\fR accepts lines on the standard
input containing \fIlatX\fR \fIlonX\fR \fIaziX\fR \fIaziY\fR \fIaziY\fR, specifying a
point where two geodesic lines \fIX\fR and \fIY\fR intersect, and prints the
location of the next closest intersection \fIx\fR \fIy\fR \fIc\fR on standard
output.
.IP 3. 4
With the \fB\-i\fR option, \fBIntersectTool\fR accepts lines on the standard
input containing \fIlatX1\fR \fIlonX1\fR \fIlatX2\fR \fIlonX2\fR \fIlatY1\fR \fIlonY1\fR
\&\fIlatY2\fR \fIlonY2\fR, specifying two geodesic segments \fIX1\fR\-\fIX2\fR and
\&\fIY1\fR\-\fIY2\fR, and prints \fIx\fR \fIy\fR \fIc\fR \fIk\fR on standard output.  Here
\&\fIk\fR is a flag in [\-4,4] specifying whether the intersection is within
the segments (0) or not (non\-zero).  \fIx\fR and \fIy\fR give the distances
from \fIX1\fR and \fIY1\fR respectively.  \fIk\fR is set to 3 \fIkx\fR + \fIky\fR
where \fIkx\fR = \-1 if \fIx\fR < 0, 0 if 0 <= \fIx\fR <= \fIsx\fR, 1 if \fIsx\fR <
\&\fIx\fR, and similarly for \fIky\fR; \fIsx\fR is the length of the segment
\&\fIX1\fR\-\fIX2\fR.
.IP 4. 4
With the \fB\-o\fR option, \fBIntersectTool\fR accepts lines on the standard
input containing \fIlatX\fR \fIlonX\fR \fIaziX\fR \fIlatY\fR \fIlonY\fR \fIaziY\fR \fIx0\fR
\&\fIy0\fR, specifying two geodesic lines \fIX\fR and \fIY\fR and two offsets,
and prints \fIx\fR \fIy\fR \fIc\fR on standard output where [\fIx\fR, \fIy\fR] is the
intersection closest to [\fIx0\fR, \fIy0\fR].
.SH OPTIONS
.IX Header "OPTIONS"
.IP \fB\-c\fR 4
.IX Item "-c"
find the closest intersection (see 1 above).
.IP \fB\-n\fR 4
.IX Item "-n"
find the intersection closest to a given intersection (see 2 above).
.IP \fB\-i\fR 4
.IX Item "-i"
find the intersection of two geodesic segments (see 3 above).
.IP \fB\-o\fR 4
.IX Item "-o"
find the closest intersection with an offset.
.IP "\fB\-R\fR \fImaxdist\fR" 4
.IX Item "-R maxdist"
modifies the four modes to return all the intersections within an L1
distance, \fImaxdist\fR, of the relevant origin: [0, 0] for \fB\-c\fR and
\&\fB\-n\fR, the midpoints of the segments for \fB\-i\fR, and [\fIx0\fR, \fIy0\fR] for
\&\fB\-o\fR.  For each intersection, \fIx\fR \fIy\fR \fIc\fR \fIz\fR is printed on
standard output.  Here \fIz\fR is the L1 distance of the intersection
from the origin and the intersections are sorted by the distances.  A
line "nan nan 0 nan" is added after the intersections, so that the
output can be associated with the correct lines of the input.  The
number of intersections scales as (\fImaxdist\fR/(pi \fIa\fR))^2.
.IP \fB\-C\fR 4
.IX Item "-C"
check the intersections.  For each computed intersection, print on
\&\fBstandard error\fR a line \fIlatX\fR \fIlonX\fR \fIlatY\fR \fIlonY\fR \fIsXY\fR giving
the computed positions of the intersections points on \fIX\fR and \fIY\fR
and the distance between them.  If \fB\-w\fR is specified, the longitude
is given before the latitude.
.IP "\fB\-e\fR \fIa\fR \fIf\fR" 4
.IX Item "-e a f"
specify the ellipsoid via the equatorial radius, \fIa\fR and
the flattening, \fIf\fR.  Setting \fIf\fR = 0 results in a sphere.  Specify
\&\fIf\fR < 0 for a prolate ellipsoid.  A simple fraction, e.g., 1/297,
is allowed for \fIf\fR.  By default, the WGS84 ellipsoid is used, \fIa\fR =
6378137 m, \fIf\fR = 1/298.257223563.
.IP \fB\-E\fR 4
.IX Item "-E"
use "exact" algorithms (based on elliptic integrals) for the geodesic
calculations.  These are more accurate than the (default) series
expansions for |\fIf\fR| > 0.02.
.IP \fB\-w\fR 4
.IX Item "-w"
on input, longitude precedes latitude (except that on input this can
be overridden by a hemisphere designator, \fIN\fR, \fIS\fR, \fIE\fR, \fIW\fR).
.IP "\fB\-p\fR \fIprec\fR" 4
.IX Item "-p prec"
set the output precision to \fIprec\fR (default 3); \fIprec\fR is the
precision relative to 1 m.  See "PRECISION".
.IP "\fB\-\-comment\-delimiter\fR \fIcommentdelim\fR" 4
.IX Item "--comment-delimiter commentdelim"
set the comment delimiter to \fIcommentdelim\fR (e.g., "#" or "//").  If
set, the input lines will be scanned for this delimiter and, if found,
the delimiter and the rest of the line will be removed prior to
processing and subsequently appended to the output line (separated by a
space).
.IP \fB\-\-version\fR 4
.IX Item "--version"
print version and exit.
.IP \fB\-h\fR 4
.IX Item "-h"
print usage and exit.
.IP \fB\-\-help\fR 4
.IX Item "--help"
print full documentation and exit.
.IP "\fB\-\-input\-file\fR \fIinfile\fR" 4
.IX Item "--input-file infile"
read input from the file \fIinfile\fR instead of from standard input; a file
name of "\-" stands for standard input.
.IP "\fB\-\-input\-string\fR \fIinstring\fR" 4
.IX Item "--input-string instring"
read input from the string \fIinstring\fR instead of from standard input.
All occurrences of the line separator character (default is a semicolon)
in \fIinstring\fR are converted to newlines before the reading begins.
.IP "\fB\-\-line\-separator\fR \fIlinesep\fR" 4
.IX Item "--line-separator linesep"
set the line separator character to \fIlinesep\fR.  By default this is a
semicolon.
.IP "\fB\-\-output\-file\fR \fIoutfile\fR" 4
.IX Item "--output-file outfile"
write output to the file \fIoutfile\fR instead of to standard output; a
file name of "\-" stands for standard output.
.SH INPUT
.IX Header "INPUT"
\&\fBIntersectTool\fR measures all angles in degrees and all lengths in
meters.  On input angles (latitude, longitude, azimuth) can be as
decimal degrees or degrees, minutes, seconds.  For example, \f(CW\*(C`40d30\*(C'\fR,
\&\f(CW\*(C`40d30\*(Aq\*(C'\fR, \f(CW\*(C`40:30\*(C'\fR, \f(CW\*(C`40.5d\*(C'\fR, and \f(CW40.5\fR are all equivalent.  By
default, latitude precedes longitude for each point (the \fB\-w\fR flag
switches this convention); however either may be given first by
appending (or prepending) \fIN\fR or \fIS\fR to the latitude and \fIE\fR or
\&\fIW\fR to the longitude.  Azimuths are measured clockwise from north;
however this may be overridden with \fIE\fR or \fIW\fR.
.PP
For details on the allowed formats for angles, see the \f(CW\*(C`GEOGRAPHIC
COORDINATES\*(C'\fR section of \fBGeoConvert\fR\|(1).
.SH PRECISION
.IX Header "PRECISION"
\&\fIprec\fR gives precision of the output with \fIprec\fR = 0 giving 1 m
precision, \fIprec\fR = 3 giving 1 mm precision, etc.  \fIprec\fR is the
number of digits after the decimal point for lengths.  The latitude
and longitude printed to standard error with the \fB\-C\fR option are
given in decimal degrees with \fIprec\fR + 5 digits after the decimal
point.  The minimum value of \fIprec\fR is 0 and the maximum is 10.
.SH ERRORS
.IX Header "ERRORS"
An illegal line of input will print an error message to standard output
beginning with \f(CW\*(C`ERROR:\*(C'\fR and causes \fBIntersectTool\fR to return an exit code
of 1.  However, an error does not cause \fBIntersectTool\fR to terminate;
following lines will be converted.
.SH ACCURACY
.IX Header "ACCURACY"
This tool will give nearly full double precision accuracy for |\fIf\fR|
< 0.02.  If the \fB\-E\fR option is given, full accuracy is achieved
for \-1/4 < f < 1/5.  The tool had not been tested outside this
range.
.SH EXAMPLES
.IX Header "EXAMPLES"
A vessel leaves Plymouth 50N 4W on a geodesic path with initial
heading 147.7W.  When will it first cross the equator?
.PP
.Vb 1
\&   echo 50N 4W 147.7W 0 0 90 | IntersectTool \-c \-p 0 \-C
\&
\&   6058049 \-3311253 0
\&   0.00000 \-29.74549 \-0.00000 \-29.74549 0
.Ve
.PP
Answer: after 6058km at longitude 29.7W.  When will it cross the date
line, longitude 180E?  Here we need to use \fB\-R\fR because there a
closer intersection on the prime meridian:
.PP
.Vb 1
\&   echo 50N 4W 147.7W 0 180 0 | IntersectTool \-c \-p 0 \-C \-R 2.6e7
\&
\&   \-494582 14052230 0 14546812
\&   53.69260 0.00000 53.69260 0.00000 0
\&   19529110 \-5932344 0 25461454
\&   \-53.51867 180.00000 \-53.51867 180.00000 0
\&   nan nan 0 nan
\&   nan nan nan nan nan
.Ve
.PP
We want the second result: after 19529 km at latitude 53.5S.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBGeoConvert\fR\|(1), \fBGeodSolve\fR\|(1).
.PP
This solution for intersections is described in C. F. F. Karney,
\&\fIGeodesic intersections\fR, J. Surveying Eng. \fB150\fR(3), 04024005:1\-9
(2024), DOI: <https://doi.org/10.1061/JSUED2.SUENG\-1483>; preprint
<https://arxiv.org/abs/2308.00495>.  It is based on the work of
S. Baseldga and J. C. Martinez\-Llario, \fIIntersection and
point\-to\-line solutions for geodesics on the ellipsoid\fR,
Stud. Geophys. Geod. \fB62\fR, 353\-363 (2018); DOI:
<https://doi.org/10.1007/s11200\-017\-1020\-z>;
.SH AUTHOR
.IX Header "AUTHOR"
\&\fBIntersectTool\fR was written by Charles Karney.
.SH HISTORY
.IX Header "HISTORY"
\&\fBIntersectTool\fR was added to GeographicLib,
<https://geographiclib.sourceforge.io>, in version 2.3.


NEWS
A reverse chronological list of changes to GeographicLib

For more information, see

    https://geographiclib.sourceforge.io/C++/doc/

The current version of the library is 2.7.

Changes between 2.7 (released 2025-11-06) and 2.6 versions:

   * Fix two bugs in classes for triaxial ellipsoids:
     + Triaxial::Geodesic3::Inverse sometimes threw an exception when
       both endpoints were on the major principal ellipse for very
       eccentric ellipsoids a > 2 c.
     + Triaxial::Cartesian3::carttocart2 sometimes returned the wrong
       result in the interior of the ellipsoid with Z = 0.

   * Cite the paper covering some of the triaxial classes Jacobi's
     solution for geodesics on a triaxial ellipsoid.

   * Minor documentation fixes.

Changes between 2.6 (released 2025-09-30) and 2.5.2 versions:

   * Support for triaxial ellipsoids. This is all placed in the
     GeographicLib::Triaxial namespace with the header files in the
     GeographicLib/Triaxial directory. This provides
     + Triaxial::Ellipsoid3 supports basic conversions of ellipsoidal
       coordinates.
     + Triaxial::GeodesicLine3 and Triaxial::Geodesic3 solve the direct
       and inverse geodesic problems on a triaxial ellipsoid. This
       implements Jacobi's solution for the direct problem. The inverse
       problem extends the method for solving the inverse problem on a
       biaxial ellipsoid.
     + Triaxial::Cartesian3 performs conversions between different
       systems of triaxial coordinates.
     + Triaxial::Conformal3 implements Jacobi's conformal projection for
       a triaxial ellipsoid. This allow a triaxial ellipsoid to be
       conformally mapped to a sphere or to any other ellipsoid.
     + Geod3Solve is a command-line utility offering access to
       Triaxial::Geodesic3.
     + Cart3Convert is a command-line utility offering access to
       Triaxial::Cartesian3.
     + Conformal3Proj is a command-line utility offering access to
       Triaxial::Conformal3.
     + Constants::Triaxial_Earth_a, Constants::Triaxial_Earth_b,
       Constants::Triaxial_Earth_c, and Constants::Triaxial_Earth_lon0
       specify a simple triaxial model for the Earth.
     + experimental::TriaxialGeodesicODE is provided to illustrate the
       solution of the direct geodesic problem by integrating the
       ordinary differential equations for the geodesic. Because this
       depends on Boost, this is not part of the GeographicLib library.
     + experimental/Geod3ODE.cpp a utility for doing direct geodesic
       calculations using experimental::TriaxialGeodesicODE.

   * To support the triaxial routines, the following classes have been
     added:
     + AngleT provides an accurate representation of angles of any
       magnitude. This might replace AuxAngle at some point.
     + Trigfun supports representing periodic functions as Fourier
       series with automatic selection of the number of terms to
       use. This offers a tiny subset of the functionality of the trig
       series in Chebfun. This might replace DST at some point.

   * C++17 is now required. Minimum version of Visual Studio supported
     is Visual Studio 15 2017.

   * Add Math::clamp.

   * Because experimental::TriaxialGeodesicODE depends on Boost, the
     CMake option USE_BOOST_FOR_EXAMPLES has been renamed
     USE_BOOST. This will cause find_package(Boost) to be called, but
     does not affect how the library or the tools are built.

Changes between 2.5.2 (released 2025-08-22) and 2.5.1 versions:

   * Fix potential buffer overflow in DMS.  Upgrading from earlier
     versions is strongly recommended.  If upgrading is not possible,
     avoid processing untrusted or malformed inputs and run programs
     linked to the library with restricted privileges.  Thanks to
     Rosario Matteo Grammatico for finding this vulnerability.

Changes between 2.5.1 (released 2025-08-20) and 2.5 versions:

   * Fix bug in geodesic inverse problem with very prolate
     ellipsoids. For f < -2, a meridional path was incorrectly
     returned in some cases for two points on the same meridian. This
     affected both the series and exact solutions. (Or course, the
     series solution is inaccurate for ellipsoids this eccentric.)

   * Minor tweak to Math::atan2d to improve accuracy.

   * Allow GEOGRAPHICLIB_PRECISION > 5 to allow use of mpreal with a
     fixed precision.

   * Document that cmake option -D BINDIR=OFF removes tools from the
     default all target for make.

   * WARNING: C++17 compliant compiler will be required for the next
     release. Support for builds with Visual Studio 14 (2015) will
     then stop.

Changes between 2.5 (released 2024-12-28) and 2.4 versions:

   * New magnetic models provided

     + World Magnetic Model 2025, wmm2025; this covers the period
       2025-2030.  This is now the model returned by
       MagneticModel::DefaultMagneticName and is default magnetic
       model for MagneticField (replacing wmm2020 which is only valid
       thru the end of 2024).
     + World Magnetic Model High Resolution 2025, wmmhr2025; this is a
       high-resolution version of wmm2025 including terms up to degree
       133 (as opposed to 12 for wmm2025).
     + International Geomagnetic Reference Field (14th generation),
       igrf14, which approximates the main magnetic field of the earth
       for the period 1900-2030.

   * Update CMake's cmake_minimum_required to 3.17.0 (to facilitate
     integration with Qt projects).

   * For GEOGRAPHICLIB_PRECISION > 2, readarray treats doubles as
     approximations to exact decimal constants.

Changes between 2.4 (released 2024-07-14) and 2.3 versions:

   * Math::sind, Math::sincosd, etc., now return accurate results for
     arguments = 30d, 45d, 60d, ...

   * Add Math::hypot3.

   * Add EllipticFunction::am.

   * Make constructor for AuxAngle explicit.

   * Include notes for plate carree areas.

   * C++14 compliant compiler required (this requirement tracks the

tools/IntersectTool.cpp
/**
 * \file IntersectTool.cpp
 * \brief Command line utility for geodesic intersections
 *
 * Copyright (c) Charles Karney (2023) <karney@alum.mit.edu> and licensed under
 * the MIT/X11 License.  For more information, see
 * https://geographiclib.sourceforge.io/
 *
 * See the <a href="IntersectTool.1.html">man page</a> for usage information.
 **********************************************************************/

#include <iostream>
#include <string>
#include <sstream>
#include <fstream>
#include <vector>
#include <GeographicLib/Geodesic.hpp>
#include <GeographicLib/DMS.hpp>
#include <GeographicLib/Utility.hpp>
#include <GeographicLib/Intersect.hpp>

#include "IntersectTool.usage"
using namespace GeographicLib;
typedef Math::real real;

int main(int argc, const char* const argv[]) {
  try {
    enum { CLOSE = 0, OFFSET, NEXT, SEGMENT };
    Utility::set_digits();
    real
      a = Constants::WGS84_a(),
      f = Constants::WGS84_f(),
      maxdist = -1;
    bool exact = false, check = false, longfirst = false;
    int prec = 3, mode = CLOSE;
    std::string istring, ifile, ofile, cdelim;
    char lsep = ';';

    for (int m = 1; m < argc; ++m) {
      std::string arg(argv[m]);
      if (arg == "-e") {
        if (m + 2 >= argc) return usage(1, true);
        try {
          a = Utility::val<real>(std::string(argv[m + 1]));
          f = Utility::fract<real>(std::string(argv[m + 2]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of -e: " << e.what() << "\n";
          return 1;
        }
        m += 2;
      } else if (arg == "-E")
        exact = true;
      else if (arg == "-p") {
        if (++m == argc) return usage(1, true);
        try {
          prec = Utility::val<int>(std::string(argv[m]));
        }
        catch (const std::exception&) {
          std::cerr << "Precision " << argv[m] << " is not a number\n";
          return 1;
        }
      } else if (arg == "-R") {
        if (++m == argc) return usage(1, true);
        try {
          maxdist = Utility::val<real>(std::string(argv[m]));
        }
        catch (const std::exception&) {
          std::cerr << "Maxdist " << argv[m] << " is not a number\n";
          return 1;
        }
        if (!(maxdist >= 0)) {
          std::cerr << "Maxdist must be nonnegative\n";
          return 1;
        }
      } else if (arg == "-c")
        mode = CLOSE;
      else if (arg == "-o")
        mode = OFFSET;
      else if (arg == "-n")
        mode = NEXT;
      else if (arg == "-i")
        mode = SEGMENT;
      else if (arg == "-C")
        check = true;
      else if (arg == "-w")
        longfirst = true;
      else if (arg == "--input-string") {
        if (++m == argc) return usage(1, true);
        istring = argv[m];
      } else if (arg == "--input-file") {
        if (++m == argc) return usage(1, true);
        ifile = argv[m];
      } else if (arg == "--output-file") {
        if (++m == argc) return usage(1, true);
        ofile = argv[m];
      } else if (arg == "--line-separator") {
        if (++m == argc) return usage(1, true);
        if (std::string(argv[m]).size() != 1) {
          std::cerr << "Line separator must be a single character\n";
          return 1;
        }
        lsep = argv[m][0];
      } else if (arg == "--comment-delimiter") {
        if (++m == argc) return usage(1, true);
        cdelim = argv[m];
      } else if (arg == "--version") {
        std::cout << argv[0] << ": GeographicLib version "
                  << GEOGRAPHICLIB_VERSION_STRING << "\n";
        return 0;
      } else
        return usage(!(arg == "-h" || arg == "--help"), arg != "--help");
    }

    if (!ifile.empty() && !istring.empty()) {
      std::cerr << "Cannot specify --input-string and --input-file together\n";
      return 1;
    }
    if (ifile == "-") ifile.clear();
    std::ifstream infile;
    std::istringstream instring;
    if (!ifile.empty()) {
      infile.open(ifile.c_str());
      if (!infile.is_open()) {
        std::cerr << "Cannot open " << ifile << " for reading\n";
        return 1;
      }
    } else if (!istring.empty()) {
      std::string::size_type m = 0;
      while (true) {
        m = istring.find(lsep, m);
        if (m == std::string::npos)
          break;
        istring[m] = '\n';
      }
      instring.str(istring);
    }
    std::istream* input = !ifile.empty() ? &infile :
      (!istring.empty() ? &instring : &std::cin);

    std::ofstream outfile;
    if (ofile == "-") ofile.clear();
    if (!ofile.empty()) {
      outfile.open(ofile.c_str());
      if (!outfile.is_open()) {
        std::cerr << "Cannot open " << ofile << " for writing\n";
        return 1;
      }
    }
    std::ostream* output = !ofile.empty() ? &outfile : &std::cout;

    Geodesic geod(a, f, exact);
    Intersect intersect(geod);
    real latX1, lonX1, aziX, latY1, lonY1, aziY, latX2, lonX2, latY2, lonY2,
      x0 = 0, y0 = 0, x, y;
    std::string inp[8], s, sc, eol;
    std::istringstream str;
    int retval = 0,
      ninp = mode == CLOSE ? 6 : (mode == NEXT ? 4 :
                                  8); // mode == OFFSET || mode == SEGMENT
    GeodesicLine lineX, lineY;
    unsigned caps = Intersect::LineCaps;
    while (std::getline(*input, s)) {
      try {
        eol = "\n";
        if (!cdelim.empty()) {
          std::string::size_type m = s.find(cdelim);
          if (m != std::string::npos) {
            eol = " " + s.substr(m) + "\n";
            s = s.substr(0, m);
          }
        }
        str.clear(); str.str(s);
        for (int i = 0; i < ninp; ++i) {
          if (!(str >> inp[i]))
          throw GeographicErr("Incomplete input: " + s);
        }
        if (str >> sc)
          throw GeographicErr("Extraneous input: " + sc);
        if (mode == CLOSE || mode == OFFSET) {
          DMS::DecodeLatLon(inp[0], inp[1], latX1, lonX1, longfirst);
          aziX = DMS::DecodeAzimuth(inp[2]);
          DMS::DecodeLatLon(inp[3], inp[4], latY1, lonY1, longfirst);
          aziY = DMS::DecodeAzimuth(inp[5]);
          if (mode == OFFSET) {
            x0 = Utility::val<real>(inp[6]);
            y0 = Utility::val<real>(inp[7]);
          } else
            x0 = y0 = 0;
          lineX = geod.Line(latX1, lonX1, aziX, caps);
          lineY = geod.Line(latY1, lonY1, aziY, caps);
        } else if (mode == NEXT) {
          DMS::DecodeLatLon(inp[0], inp[1], latX1, lonX1, longfirst);
          aziX = DMS::DecodeAzimuth(inp[2]);
          aziY = DMS::DecodeAzimuth(inp[3]);
          lineX = geod.Line(latX1, lonX1, aziX, caps);
          lineY = geod.Line(latX1, lonX1, aziY, caps);
        } else {                // mode == SEGMENT
          DMS::DecodeLatLon(inp[0], inp[1], latX1, lonX1, longfirst);
          DMS::DecodeLatLon(inp[2], inp[3], latX2, lonX2, longfirst);
          DMS::DecodeLatLon(inp[4], inp[5], latY1, lonY1, longfirst);
          DMS::DecodeLatLon(inp[6], inp[7], latY2, lonY2, longfirst);
          lineX = geod.InverseLine(latX1, lonX1, latX2, lonX2,
                                   Intersect::LineCaps);
          lineY = geod.InverseLine(latY1, lonY1, latY2, lonY2,
                                   Intersect::LineCaps);
          x0 = lineX.Distance()/2;
          y0 = lineY.Distance()/2;
        }
        std::pair<real, real> p0(x0, y0);
        if (maxdist < 0) {
          int segmode = 0, c;
          auto p = mode == CLOSE || mode == OFFSET ?
            intersect.Closest(lineX, lineY, p0, &c) :
            mode == NEXT ? intersect.Next(lineX, lineY, &c) :
            intersect.Segment(lineX, lineY, segmode, &c);
          x = p.first; y = p.second;
          *output << Utility::str(x, prec) << " "
                  << Utility::str(y, prec) << " " << c;
          if (mode == SEGMENT)
            *output << " " << segmode;
          *output << eol;
          if (check) {
            lineX.Position(x, latX2, lonX2);
            lineY.Position(y, latY2, lonY2);
            std::cerr << Utility::str(latX2, prec + 5) << " "
                      << Utility::str(lonX2, prec + 5) << " "
                      << Utility::str(latY2, prec + 5) << " "
                      << Utility::str(lonY2, prec + 5) << eol;
          }
        } else {
          std::vector<int> c;
          auto v = intersect.All(lineX, lineY, maxdist, c, p0);
          unsigned n = unsigned(v.size());
          for (unsigned i = 0; i < n; ++i) {
            x = v[i].first; y = v[i].second;
            *output << Utility::str(x, prec) << " " << Utility::str(y, prec)
                    << " " << c[i] << " "
                    << Utility::str(Intersect::Dist(v[i], p0), prec)
                    << eol;
            if (check) {
              lineX.Position(x, latX2, lonX2);
              lineY.Position(y, latY2, lonY2);
              std::cerr << Utility::str(latX2, prec + 5) << " "
                        << Utility::str(lonX2, prec + 5) << " "
                        << Utility::str(latY2, prec + 5) << " "
                        << Utility::str(lonY2, prec + 5) << eol;
            }
          }
          *output << "nan nan 0 nan" << eol;
          if (check)
            std::cerr << "nan nan nan nan" << eol;
        }
      }
      catch (const std::exception& e) {
        // Write error message cout so output lines match input lines
        *output << "ERROR: " << e.what() << "\n";
        retval = 1;
      }
    }
    return retval;
  }
  catch (const std::exception& e) {
    std::cerr << "Caught exception: " << e.what() << "\n";
    return 1;
  }
  catch (...) {
    std::cerr << "Caught unknown exception\n";
    return 1;
  }
}
